### ISSUE TITLE
[TF2] [Linux] Inconsistent crashes and/or segfaults roughly every 45 minutes, possibly unrelated VAC warnings

### ISSUE THREAD

**Issue by superselect** [ValveSoftware/Source-1-Games#4934]
For the past month or so, TF2 has been constantly crashing. When it first started happening, it would crash within seconds of loading into a map. I did a clean reinstall, which helped, but the crashes still happened. It seems to do it 20 minutes to two hours after launching, with a significant amount occurring during a level transition. I don't find restarting the game super annoying on it's own, so I just ignored it. The VAC warnings are what made me finally report this.

It always crashes in one of three ways. The game will either crash without any warning whatsoever, or will show a popup saying either "failed to lock vertex buffer in CMeshDX8::LockVertexBuffer: nVertexCount=49142, nFirstVertex=0" or "Out of memory or address space. Texture quality setting may be too high." When it crashes with a popup, a segfault is logged. See the syslog for details. There is no relevant info in console.log, except a line containing the message shown in the popup.

A day or two ago, after restarting tf2 after a crash, I joined a server, and got a warning from VAC linking to [this article](https://help.steampowered.com/en/faqs/view/22C0-03D0-AE4B-04E8). That same sequence happened again today with nothing in tf/custom (I had mastercomfig before). The only other programs I had running both times were firefox and likely libvirt (no VM running, just the libvirt service). Both times, Steam successfully verified all files after. But the first time time, after updating my computer, rebooting, starting the game again, going into a local server with the `map` command, noticing some weird errors in the console, and closing the game... two files failed verification.

I have no clue what's going on at this point. It could feasibly be multiple unrelated issues.
It's also likely that something is built or configured wrong in my OS, but I'm not finding anything useful in the logs to indicate what.

# Logs and info

[System info](https://gist.github.com/superselect/ba4c202c6c8e44babbf850a0bc745560)

[Console.log](https://gist.github.com/superselect/ff60840b446b87666a3b0c6314c2b92b) right after the second warning from VAC. The log starts at a level transition. The game crashed during that level transition with no popup or warning. It also includes starting the game again, joining the server, and getting the VAC warning.

[Relevant parts of syslog](https://gist.github.com/superselect/8691dd9183b27bb7a7035f01488bfc87). Should correspond with many of the dumps below.

Contents of /tmp/dumps/
[dumps_just_after_first_vac_warning.zip](https://github.com/ValveSoftware/Source-1-Games/files/11364300/dumps_just_after_first_vac_warning.zip)
[dumps_just_after_second_vac_warning.zip](https://github.com/ValveSoftware/Source-1-Games/files/11364301/dumps_just_after_second_vac_warning.zip)

**Comment by Pants** [comment]
I'm also experiencing this issue, it seems to be related to Intel GPUs. I noticed OP has one in their system info, and this started to happen once I installed mine.

**Comment by superselect** [comment]
Interesting, is it one of the new dedicated GPUs? I'm just using integrated graphics. Including your system info might help.

I've done some additional testing, and, on the same hardware, tf2 also crashes on a fully updated Arch Linux install (as of a week or so ago), but it does not crash on a fresh Linux Mint install (as of a few days ago).

[Arch Linux system info, crashing](https://gist.github.com/superselect/89c39f2a4fbb01f4928ef29445cf392a)
[Arch Linux (LTS kernel) system info, crashing](https://gist.github.com/superselect/99e7bc0be34acdf5b3628f671ed4c73e)

[Linux Mint system info, not crashing](https://gist.github.com/superselect/7f66a05aa219e0cf01a6869c150fe2f2)

At the moment, I'm leaning towards it being an issue in mesa, but that's mostly just speculation based on the fact that the crash *sometimes* happens in iris_dri.so. I might try downgrading it later.

**Comment by superselect** [comment]
Yep, installing mesa-22.2.5 fixed it. Closing as this does not appear to be an issue with tf2.

**Comment by superselect** [comment]
Installing mesa-22.2.5 did not fix the problem. But it does make it usable. Every time I install the most recent version of mesa, the game won't last 2 seconds at the main menu before dying. It may be relevant that mesa-22.2.5 needs llvm 15, while newer versions are compiled against llvm 16 on my system.

I'm reopening this because I still have no clue what's going on, and the problem could realistically be in the engine.

The problem manifests in so many different ways. Currently, maybe half the time the game will crash on startup. If I just play on servers, it will crash randomly after at least an hour or so with various errors. Starting the game, connecting to a server, disconnecting, and loading a map with the internal server crashes it very consistently at the moment, but just loading a map after starting the game doesn't crash it. All of that can change very easily though.
I've noticed that changes in behavior correlate with the "compiling vulkan shaders" box showing up.

This doesn't seem exclusive to tf2, but I can't make any definite statements about that. It is absolutely exclusive to Source (and Source 2) based games though. Even CS2 appears to be doing something similar.

I just saw [this](https://github.com/ValveSoftware/Source-1-Games/issues/5126#issuecomment-1675782080) and decided to try adding -heapsize 2097152 to the launch options. It doesn't fix the issue, but it does help significantly. The game still crashes randomly on launch, but when it gets past that point, it lasts far longer than it did before. It does not help with the most recent mesa, the game still crashes within seconds of getting to the main menu.
